File: apps/ios/ExpCore/Sources/Domain/RunResume.swift
import Foundation

/// EXP-637: whether an ended run can be picked up again, and on which machine.
///
/// Resume relaunches the pinned agent in the run's OWN worktree, so it is
/// bound to the machine that ran it: `steer.startSession({ resumeSessionId })`
/// refuses a foreign session (owner-only), a still-live one, another machine,
/// and one whose device does not advertise `resume-run`. The clients mirror
/// those gates so the affordance never appears where the send would bounce.
public enum RunResume {
    /// The machine a Resume would go to, or nil when this run cannot be
    /// resumed from here. Prefers the caller's OWN devices row when a shared
    /// teammate row happens to carry the same machine id.
    public static func target(
        for session: CodingSessionEntity,
        devices: [SteerDevice],
        currentUserId: String?
    ) -> SteerDevice? {
        // EXP-888: a sweep end (`ended_by == "stale"`) is NOT an end — the
        // host ignores the flip and heartbeats the row back to `running`.
        // Resuming it would put a second agent on the run's worktree.
        guard CodingSessionOwnership.isOwn(session, userId: currentUserId),
              PastRuns.hasEnded(session),
              let deviceId = session.deviceId, !deviceId.isEmpty
        else { return nil }
        let candidates = devices.filter { $0.deviceId == deviceId }
        let device = candidates.first(where: \.isMine) ?? candidates.first
        guard let device, device.isOnline, device.canResumeRun else { return nil }
        return device
    }
}

File: apps/android/app/src/main/java/com/exponential/app/domain/RunResume.kt
package com.exponential.app.domain

import com.exponential.app.data.api.SteerDevice
import com.exponential.app.data.db.CodingSessionEntity

// EXP-637: when a FINISHED coding session can be resumed, and where.
//
// An agent closes its own run out through the `exponential_sessions_end` MCP
// tool: a one-paragraph summary. EXP-686 dropped the self-reported outcome
// everywhere — an ended run shows its name, when it ended and that summary,
// nothing else.

/**
 * Where a Resume would go: the machine that ran the session, resolved against
 * the live device rows.
 */
data class RunResumeTarget(
    val sessionId: String,
    val deviceId: String,
    val deviceLabel: String,
)

/**
 * Whether [session] can be resumed right now, and on which machine (EXP-637).
 *
 * Everything the server checks, checked here first so the affordance is simply
 * absent instead of failing after the tap: the run must be the caller's OWN (a
 * teammate's run is never resumable — EXP-312), it must have ENDED (a live run
 * is steered, not resumed), it must carry the `device_id` of the machine that
 * ran it (that machine still holds the worktree and the agent's transcript),
 * and that machine must be online AND advertise the `resume-run` cap. The
 * machine may be a teammate's SHARED server (EXP-432) — the run is still the
 * caller's; own rows just win when the same machine id appears twice.
 *
 * Mirrors iOS `RunResume.target(for:devices:currentUserId:)`.
 */
fun resumeTargetFor(
    session: CodingSessionEntity,
    devices: List<SteerDevice>,
    currentUserId: String?,
): RunResumeTarget? {
    if (currentUserId == null || session.userId != currentUserId) return null
    if (session.status != DomainContract.codingSessionStatusEnded) return null
    val deviceId = session.deviceId?.takeIf { it.isNotEmpty() } ?: return null
    val candidates = devices.filter { it.deviceId == deviceId }
    val device = candidates.firstOrNull { it.isMine } ?: candidates.firstOrNull() ?: return null
    if (!device.online || !device.canResumeRun) return null
    return RunResumeTarget(
        sessionId = session.id,
        deviceId = device.deviceId,
        // Never blank: an unlabelled machine still reads as something.
        deviceLabel = device.deviceLabel.ifBlank { device.deviceId },
    )
}
